utils/land-patterns.stanza

```diff
@@ -71,270 +168,260 @@ public defn soldermask-defined (p:Pad):
+  val d = dims(hole)
+  layer(Courtyard(Top))    = Rectangle(x(d), y(d))
+  layer(Courtyard(Bottom)) = Rectangle(x(d), y(d))
+
+; Create a non-plated hole landpattern with a hole shape
+public defn non-plated-hole-landpattern (hole:Shape):
+  non-plated-hole-landpattern(hole, offset(hole,  SOLDER-MASK-REGISTRATION))
+
+; Create a circular non-plated hole landpattern with given drill radius and mask radius
+public defn non-plated-hole-landpattern (drill-r:Double, mask-r:Double) :
+  non-plated-hole-landpattern(Circle(drill-r), Circle(mask-r))
+
+; Create a non-plated hole landpattern with a given drill radius.
+public defn non-plated-hole-landpattern (drill-r:Double) :
+  non-plated-hole-landpattern(drill-r, drill-r)
+
+;==============================================================================
+;========================== Land Pattern Generators ===========================
+;==============================================================================
+
+;==============================================================================
+;========================== Pose Arrangement Helpers ==========================
+;==============================================================================
+; PinPadMapping is a not-yet-placed pad, with assigned name, pad, and pose
+pcb-struct ocdb/land-patterns/PinPadMapping:
+  pad-name:  Ref,  ; the name of the pad
+  pad-shape: JITXDef, ; the corresponding pcb-pad
+  pose:      Pose,    ; the pose of the pad
+
+; A generic landpattern created from a sequence of PinPadMappings
+pcb-landpattern make-generic-landpattern (pads:Seqable<PinPadMapping>):
   inside pcb-landpattern:
-    
-    val n-col = to-int(ceil(to-double(n-pin) / to-double(n-row)))
-
-    val l0 = loc((to-double(n-row - 1) * row-pitch) / -2.0, (to-double(n-col - 1) * pin-pitch) / 2.0)
-    
-    for i in 0 to n-col do :
-      for j in 0 to n-row do :
-        pad p[n-row * i + j + 1] : pad-ref at l0 * loc(to-double(j) * row-pitch, (- to-double(i) * pin-pitch))
-    
-public defn make-pin-grid (n-pin:Int, n-row:Int, pin-pitch:Double, row-pitch:Double) :
-  make-pin-grid(n-pin, n-row, pin-pitch, row-pitch, pth-pad(0.5))
-
-public defn make-pin-grid (n-pin:Int) :
-  make-pin-grid(n-pin, 1, 2.54, 2.54)
-
+    for p in pads do:
+      pad (pad-name(p)): (pad-shape(p) as Pad) at pose(p)
+
+; Create a generic landpattern with a uniform pad
+defn make-generic-landpattern (pad-names:Seqable<Ref>,
+                               poses:Seqable<Pose>,
+                               pad-shape:JITXDef):
+  val pads = 
+    for (pn in pad-names, pp in poses) seq:
+      PinPadMapping(pn, pad-shape, pp) 
+  make-generic-landpattern(pads)
+
+; Naming for a 2-column arrangment of pins 
+public defn soic-naming-convention (row:Int, column:Int, num-pins:Int) -> Ref:
+  if column == 0:
+    IndexRef(Ref("p"), row + 1)
+  else if column == 1:
+    IndexRef(Ref("p"), num-pins - row)
+  else:
+    fatal("soic naming convention supports a maximum of 2 columns")
```

Comment:
It would prefer having at the beginning of the function, such that it could be clearly seen as a precondition check